apiserver/common/secrets/secrets.go

```diff
@@ -672,87 +672,136 @@ func RemoveSecretsForAgent(
-		}
-		if len(revisions) == 0 {
-			// Remove all revisions.
-			revs, err := removeState.ListSecretRevisions(uri)
+		// Repeatedly attempt to delete the revision from the backend until one of:
+		// * deletion is successful
+		// * deletion is unsuccessful with error==NotFound but the revision has not been moved to a new backend
+		// * deletion is unsuccessful with error!=NotFound
+		for {
+			cfgInfo, err := adminConfigGetter()
 			if err != nil {
 				return errors.Trace(err)
 			}
-			for _, rev := range revs {
-				gatherExternalRevs(rev.ValueRef)
-			}
-		} else {
-			for _, rev := range revisions {
-				revMeta, err := removeState.GetSecretRevision(uri, rev)
-				if err != nil {
-					return errors.Trace(err)
-				}
-				gatherExternalRevs(revMeta.ValueRef)
+			backendCfg := cfgInfo.Configs[backendId]
+			err = removeRevisionFromBackend(&backendCfg, revisionId)
+			if err == nil {
+				// Continue to the next revision if successful.
+				break
+			} else if !errors.Is(err, errors.NotFound) {
+				// Exit early for any error other than NotFound
+				return errors.Trace(err)
 			}
-		}
 
-		for backendID, r := range externalRevs {
-			backendCfg, ok := cfgInfo.Configs[backendID]
-			if !ok {
-				return errors.NotFoundf("secret backend %q", backendID)
-			}
-			provider, err := GetProvider(backendCfg.BackendType)
+			// NotFound could be because:
+			// 1. The backend is draining and the secret was moved to the new backend before we accessed it.
+			// 2. The secret is actually missing from the backend.
+			// Check if the revision has moved to a new backend.
+			updatedRev, err := removeState.GetSecretRevision(uri, rev.Revision)
 			if err != nil {
 				return errors.Trace(err)
 			}
-			if err := removeFromBackend(provider, backendCfg, r); err != nil {
-				return errors.Trace(err)
+
+			// If the backend changed, try to delete the secret from the new backend.
+			if backendId != updatedRev.ValueRef.BackendID {
+				backendId = updatedRev.ValueRef.BackendID
+				revisionId = updatedRev.ValueRef.RevisionID
+				continue
 			}
+
+			// Otherwise, the revision really is missing from the backend and we move on.
+			// We tolerate this because our goal is to have that revision removed anyway.
+			break
 		}
-		return nil
+
+		// Log this revision to be cleaned up in the provider
+		if _, ok := providersToCleanUp[backendId]; !ok {
+			providersToCleanUp[backendId] = provider.SecretRevisions{}
+		}
+		providersToCleanUp[backendId].Add(uri, rev.ValueRef.RevisionID)
+	}
+
+	// Clean up all providers we've touched
+	for backendId, secretRevisions := range providersToCleanUp {
+		cfgInfo, err := adminConfigGetter()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		backendCfg := cfgInfo.Configs[backendId]
+		p, err := GetProvider(cfgInfo.Configs[backendId].BackendType)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if err := p.CleanupSecrets(&backendCfg, authTag, secretRevisions); err != nil {
```

Comment:
Should this only be done on backends we successfully delete from?  Ex:
* we try to delete rev1 from backend1 and its NotFound
* we see it moved to backend2, so we try and succeed in deleting it in backend2
* should we CleanupSecrets on backend1 and backend2, or just backend2?

My first guess is just cleanup backend2, since whoever moved the secret out of backend1 should have done the cleanup itself.

Reply:
That's correct, if a secret has been drained the associated ACLs should also have been dealt with. Wouldn't hurt to double check.